components/callbacks/executors_test.go

```diff
@@ -223,6 +237,168 @@ func TestProcessInvocationTask_Outcomes(t *testing.T) {
+			timer.EXPECT().Record(gomock.Any(),
+				metrics.NamespaceTag("namespace-name"),
+				metrics.DestinationTag("http://localhost"),
+				metrics.StringTag("hsm-callback", tc.expectedMetricOutcome))
+
+			root := newRoot(t)
+			ref := &persistencespb.StateMachineRef{
+				Path: []*persistencespb.StateMachineKey{
+					{
+						Type: scheduler.WorkflowType,
+						Id:   "testId",
+					},
+				},
+			}
+			cb := callbacks.Callback{
+				CallbackInfo: &persistencespb.CallbackInfo{
+					Callback: &persistencespb.Callback{
+						Variant: &persistencespb.Callback_Hsm_{
+							Hsm: &persistencespb.Callback_Hsm{
+								NamespaceId: "nsid",
+								WorkflowId:  "wid",
+								RunId:       "rid",
+								Ref:         ref,
+								Method:      "test",
+							},
+						},
+					},
+					State: enumsspb.CALLBACK_STATE_SCHEDULED,
+				},
+			}
+			coll := callbacks.MachineCollection(root)
+			node, err := coll.Add("ID", cb)
+			require.NoError(t, err)
+			env := fakeEnv{node}
+
+			key := definition.NewWorkflowKey("namespace-id", "", "")
+			reg := hsm.NewRegistry()
+			historyClientMock := historyservicemock.NewMockHistoryServiceClient(ctrl)
+
+			historyClientMock.EXPECT().InvokeStateMachineMethod(gomock.Any(), gomock.Any(), gomock.Any()).DoAndReturn(func(ctx context.Context, in *historyservice.InvokeStateMachineMethodRequest, opts ...grpc.CallOption) (*historyservice.InvokeStateMachineMethodResponse, error) {
+				require.Equal(t, "nsid", in.NamespaceId)
+				require.Equal(t, "wid", in.WorkflowId)
+				require.Equal(t, "rid", in.RunId)
+				require.True(t, ref.Equal(in.Ref))
+				require.Equal(t, "test", in.MethodName)
+				event := &historypb.HistoryEvent{}
+				err = proto.Unmarshal(in.Input, event)
+				require.NoError(t, err)
+				require.Equal(t, int64(42), event.EventId)
+				require.Equal(t, enums.EVENT_TYPE_WORKFLOW_EXECUTION_COMPLETED, event.EventType)
+
+				return &historyservice.InvokeStateMachineMethodResponse{}, tc.expectedError
+			}).Times(1)
+			require.NoError(t, callbacks.RegisterExecutor(
+				reg,
+				callbacks.TaskExecutorOptions{
+					NamespaceRegistry: namespaceRegistryMock,
+					MetricsHandler:    metricsHandler,
+					HistoryClient:     historyClientMock,
+					Logger:            log.NewNoopLogger(),
+					Config: &callbacks.Config{
+						RequestTimeout: dynamicconfig.GetDurationPropertyFnFilteredByDestination(time.Second),
+						RetryPolicy: func() backoff.RetryPolicy {
+							return backoff.NewExponentialRetryPolicy(time.Second)
+						},
+					},
+				},
+			))
+
+			err = reg.ExecuteImmediateTask(
+				context.Background(),
+				env,
+				hsm.Ref{
+					WorkflowKey: key,
+					StateMachineRef: &persistencespb.StateMachineRef{
+						Path: []*persistencespb.StateMachineKey{
+							{
+								Type: callbacks.StateMachineType,
+								Id:   "ID",
+							},
+						},
+					},
+				},
+				callbacks.InvocationTask{Destination: "http://localhost"},
```

Comment:
Destination should be empty for HSM callbacks, destination is only for outbound tasks.
We need to also add support for executing HSM tasks on the transfer queue before we finalize this work.
I can take that on next week.